src/lightning/data/processing/readers.py

```diff
@@ -0,0 +1,120 @@
+import os
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, List, Optional
+
+from lightning_utilities.core.imports import RequirementCache
+
+from lightning.data.streaming.shuffle import _associate_chunks_and_internals_to_ranks
+from lightning.data.utilities.env import _DistributedEnv
+
+_POLARS_AVAILABLE = RequirementCache("polars")
+_PYARROW_AVAILABLE = RequirementCache("pyarrow")
+
+
+class BaseReader(ABC):
+
+    def get_num_nodes(self) -> int:
+        return int(os.getenv("DATA_OPTIMIZER_NUM_NODES", 1))
+
+    @abstractmethod
+    def to_workers_user_items(self, items: List[Any], num_workers: int) -> List[List[Any]]:
```

Comment:
Not sure `user` is what we want for the name. Maybe `shard_items` or `items_to_workers`?

Reply:
done !